### PR TITLE
Restapi menu

### PR DESCRIPTION
**Pull request by PNH00**

### CODE REVIEW

**Comment by PNH00** [comment]
[Implement RESTful API for CRUD operation to manage menu in restaurant](https://trello.com/c/l9kkvJmy/15-implement-restful-api-for-crud-operation-to-manage-menu-in-restaurant)

File: RestaurentManagement/src/main/java/com/restapi/controllers/MenuController.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    public ResponseEntity<List<Menu>> getAllMenus() {
+        List<Menu> menus = menuService.getAllMenus();
+        return new ResponseEntity<>(menus, HttpStatus.OK);
+    }
+
+    @GetMapping("/{id}")
+    public ResponseEntity<?> getMenuById(@PathVariable UUID id) {
+        try {
+            Optional<Menu> menuOptional = menuService.getMenuById(id);
+            Menu menu = menuOptional.orElseThrow(() -> new ResourceException(menuService.createNotFoundErrorDetail(id)));
```
Comment (vtronghieu) [inline RestaurentManagement/src/main/java/com/restapi/controllers/MenuController.java]: Pls move the exception handle to the service

File: RestaurentManagement/src/main/java/com/restapi/controllers/MenuController.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            return new ResponseEntity<>(errorDetail, HttpStatus.valueOf(errorDetail.getCode()));
+        }
+    }
+
+    @PutMapping("/{id}")
+    public ResponseEntity<?> updateMenu(@PathVariable UUID id, @RequestBody Menu menu) {
+        try {
+            Menu updatedMenu = menuService.updateMenu(id, menu);
+            return new ResponseEntity<>(updatedMenu, HttpStatus.OK);
+        } catch (ResourceException e) {
```
Comment (vtronghieu) [inline RestaurentManagement/src/main/java/com/restapi/controllers/MenuController.java]: This exception, too

File: RestaurentManagement/src/main/java/com/restapi/controllers/MenuController.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            return new ResponseEntity<>(errorDetail, HttpStatus.valueOf(errorDetail.getCode()));
+        }
+    }
+
+    @DeleteMapping("/{id}")
+    public ResponseEntity<ErrorDetail> deleteMenu(@PathVariable UUID id) {
+        try {
+            menuService.deleteMenu(id);
+            return new ResponseEntity<>(HttpStatus.NO_CONTENT);
+        } catch (ResourceException e) {
```
Comment (vtronghieu) [inline RestaurentManagement/src/main/java/com/restapi/controllers/MenuController.java]: Pls re-check other exceptions

File: RestaurentManagement/src/main/java/com/restapi/exceptions/ErrorDetail.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+import lombok.Data;
+
+@Data
+public class ErrorDetail {
+    private String timestamp;
+    private String status;
+    private int code;
+    private String message;
+    private Object data;
```
Comment (vtronghieu) [inline RestaurentManagement/src/main/java/com/restapi/exceptions/ErrorDetail.java]: Don't show any data when error the API

File: RestaurentManagement/src/main/java/com/restapi/service/MenuService.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    private final MenuRepository menuRepository;
+    private final TypeRepository typeRepository;
+    @Autowired
+    public MenuService(MenuRepository menuRepository, TypeRepository typeRepository) {
+        this.menuRepository = menuRepository;
+        this.typeRepository = typeRepository;
+    }
+
+    public ErrorDetail createNotFoundErrorDetail(UUID id) {
```
Comment (vtronghieu) [inline RestaurentManagement/src/main/java/com/restapi/service/MenuService.java]: Move the handle of the error detail to another class to handle it.

File: RestaurentManagement/src/main/java/com/restapi/service/TypeService.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+public class TypeService {
+
+    private final TypeRepository typeRepository;
+
+    @Autowired
+    public TypeService(TypeRepository typeRepository) {
+        this.typeRepository = typeRepository;
+    }
+
+    public ErrorDetail createNotFoundErrorDetail(UUID id) {
```
Comment (vtronghieu) [inline RestaurentManagement/src/main/java/com/restapi/service/TypeService.java]: Move this handle, too

File: RestaurentManagement/src/main/java/com/restapi/exceptions/ErrorDetailHandler.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
+package com.restapi.exceptions;
+
+import com.restapi.models.Menu;
+import org.springframework.http.HttpStatus;
+import java.util.Date;
+import java.util.UUID;
+
+public class ErrorDetailHandler {
+    public ErrorDetail createNotFoundErrorDetail(UUID id) {
```
Comment (vtronghieu) [inline RestaurentManagement/src/main/java/com/restapi/exceptions/ErrorDetailHandler.java]: Pls move the body code in the functions in GlobalExceptionHandler

File: RestaurentManagement/src/main/java/com/restapi/exceptions/ErrorDetailHandler.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        ErrorDetail errorDetail = new ErrorDetail();
+        errorDetail.setTimestamp(new Date().toString());
+        errorDetail.setStatus("Not found");
+        errorDetail.setCode(HttpStatus.NOT_FOUND.value());
+        errorDetail.setMessage("Check your id '" + id + "' and try again");
+        return errorDetail;
+    }
+
+    public ErrorDetail createNotFoundErrorDetail(String message, Menu menu) {
```
Comment (vtronghieu) [inline RestaurentManagement/src/main/java/com/restapi/exceptions/ErrorDetailHandler.java]: This body code, too

File: RestaurentManagement/src/main/java/com/restapi/service/MenuService.java [hunk 9]
```diff
@@ -1,50 +1,35 @@
...
 import org.springframework.stereotype.Service;
+import org.springframework.web.server.ResponseStatusException;
 import java.util.*;
 
 @Service
 public class MenuService {
 
     private final MenuRepository menuRepository;
     private final TypeRepository typeRepository;
+    private final ErrorDetailHandler errorDetailHandler = new ErrorDetailHandler();
```
Comment (vtronghieu) [inline RestaurentManagement/src/main/java/com/restapi/service/MenuService.java]: Don't need to call error object in service classes

File: RestaurentManagement/src/main/java/com/restapi/service/MenuService.java [hunk 10]
```diff
@@ -1,50 +1,35 @@
...
-        errorDetail.setCode(HttpStatus.BAD_REQUEST.value());
-        errorDetail.setMessage(message);
-        errorDetail.setData(menu);
-        return errorDetail;
-    }
-
     public Menu createMenu(Menu menu) {
         if (menu.getType().isEmpty()) {
-            ErrorDetail errorDetail = createNotFoundErrorDetail("Cannot create!",menu);
+            ErrorDetail errorDetail = errorDetailHandler.createNotFoundErrorDetail("Cannot create!",menu);
```
Comment (vtronghieu) [inline RestaurentManagement/src/main/java/com/restapi/service/MenuService.java:32]: Pls use "throw new your-exception"

Comment (vtronghieu) [inline RestaurentManagement/src/main/java/com/restapi/service/MenuService.java]: After modifying the comment before, you don't need use this code line

File: RestaurentManagement/src/main/java/com/restapi/service/MenuService.java [hunk 11]
```diff
@@ -66,21 +51,13 @@ public List<Menu> getAllMenus() {
     }
 
     public Optional<Menu> getMenuById(UUID id){
-        if (!menuRepository.existsById(id)) {
-            ErrorDetail errorDetail = createNotFoundErrorDetail(id);
-            throw new ResourceException(errorDetail);
-        }
         return menuRepository.findById(id);
```
Comment (vtronghieu) [inline RestaurentManagement/src/main/java/com/restapi/service/MenuService.java:59]: Pls use "throw new your-exception" for not found id

File: RestaurentManagement/src/main/java/com/restapi/service/MenuService.java [hunk 12]
```diff
@@ -91,10 +68,17 @@ public Menu updateMenu(UUID id, Menu menu) {
...
     public void deleteMenu(UUID id) {
-        if (!menuRepository.existsById(id)) {
-            ErrorDetail errorDetail = createNotFoundErrorDetail(id);
-            throw new ResourceException(errorDetail);
+        try {
+            if (!menuRepository.existsById(id)) {
+                throw new EmptyResultDataAccessException(1);
+            }
+            menuRepository.deleteById(id);
+        } catch (EmptyResultDataAccessException e) {
```
Comment (vtronghieu) [inline RestaurentManagement/src/main/java/com/restapi/service/MenuService.java]: Just need to check the existing ID and throw an exception if not found

File: RestaurentManagement/src/main/java/com/restapi/service/MenuService.java [hunk 13]
```diff
@@ -91,10 +68,17 @@ public Menu updateMenu(UUID id, Menu menu) {
...
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Menu not found with id: " + id);
+        } catch (DataAccessException e) {
```
Comment (vtronghieu) [inline RestaurentManagement/src/main/java/com/restapi/service/MenuService.java]: If u want to ensure this API, just need to catch (Exception e) enough

File: RestaurentManagement/src/main/java/com/restapi/service/TypeService.java [hunk 14]
```diff
@@ -17,29 +16,21 @@
 public class TypeService {
 
     private final TypeRepository typeRepository;
+    @Getter
+    private final ErrorDetailHandler errorDetailHandler = new ErrorDetailHandler();
```
Comment (vtronghieu) [inline RestaurentManagement/src/main/java/com/restapi/service/TypeService.java]: Modify follow the comments before

**Pull request merged**